juju/application.py

```diff
@@ -222,30 +228,129 @@ async def destroy(self):
         """Remove this application from the model.
 
         """
-        app_facade = client.ApplicationFacade.from_connection(self.connection)
+        app_facade = self._facade()
 
         log.debug(
             'Destroying %s', self.name)
 
         return await app_facade.Destroy(application=self.name)
     remove = destroy
 
-    async def expose(self):
-        """Make this application publicly available over the network.
+    async def expose(self, exposed_endpoints=None):
+        """Make a subset of the application endpoints or the entire application
+           available over the network.
+
+           If the exposed_endpoints argument is not provided, all opened port
+           ranges for the application will become reachable from 0.0.0.0/0.
+
+           On juju 2.9 and onwards, the exposed_endpoints argument may be used
+           to specify a list of spaces and or CIDRs that should be able to
+           reach the port ranges opened for a particular subnet. The
+           exposed_endpoints parameter is a map where keys are endpoint names
+           or the empty string ("") which works as a wildcard for all endpoints
+           and values are ExposedEndpoint instances.
+
+           When targeting an older juju controller, the exposed_endpoints param
+           is not supported and an error will be raised if it is provided.
+        """
+        app_facade = self._facade()
+        facade_version = self._facade_version()
+
+        if exposed_endpoints is not None:
+            if not isinstance(exposed_endpoints, dict):
+                raise ValueError("endpoints must be a dictionary with ExposedEndpoint values")
+
+            # The bundle changes code will pass in raw dicts with the exposed
+            # endpoint data. We need to convert those into ExposedEndpoints
+            for k, v in exposed_endpoints.items():
+                if not isinstance(v, ExposedEndpoint):
+                    exposed_endpoints[k] = ExposedEndpoint.from_dict(v)
+
+            # Check if the specified exposed_endpoints would cause security
+            # issues when applied to a pre 2.9 controller.
```

Comment:
Is this done client-side in Juju CLI as well?

Reply:
Yes, both for [expose](https://github.com/juju/juju/blob/develop/api/application/client.go#L865) and [unexpose](https://github.com/juju/juju/blob/develop/api/application/client.go#L902). The contents of the error messages also match so we are consistent.